File: common/src/main/java/cn/breadnicecat/candycraftce/utils/tools/Accessor.java
package cn.breadnicecat.candycraftce.utils.tools;

import java.util.function.Consumer;
import java.util.function.Supplier;

/**
 * @author <a href="https://gitee.com/Bread_NiceCat">Bread_NiceCat</a>
 * @date 2023/1/22 13:21
 * <p>
 * Wisdom Util ;)
 */
public interface Accessor<T> extends Consumer<T>, Supplier<T> {
	
	@Override
	@Deprecated
	default void accept(T t) {
		set(t);
	}
	
	void set(T t);
	
	default Accessor<T> frozen() {
		return new Frozen<>(this);
	}
	
	class Frozen<T> implements Accessor<T> {
		private final Accessor<T> su;
		
		public Frozen(Accessor<T> su) {
			this.su = su;
		}
		
		@Override
		public T get() {
			return su.get();
		}
		
		@Override
		public void set(T t) {
			throw new UnsupportedOperationException("frozen");
		}
		
		@Override
		public Accessor<T> frozen() {
			return this;
		}
	}
}

File: src/main/kotlin/cn/breadnicecat/candycraftce/utils/Accessor.kt
package cn.breadnicecat.candycraftce.utils

import java.util.function.Consumer
import java.util.function.Supplier

/**
 * @author [Bread_NiceCat](https://gitee.com/Bread_NiceCat)
 * @date 2023/1/22 13:21
 *
 * Wisdom Util ;)
 */
sealed interface Accessor<T> : Consumer<T>, Supplier<T> {
    @Deprecated("Use `set` instead", ReplaceWith("set"))
    override fun accept(t: T) = set(t)

    fun set(t: T)
    fun frozen(): Accessor<T> = Frozen(this)
    fun synchronized(): Accessor<T> = SynchronizedAccessor(this)

    companion object {
        fun <T> access(getter: Supplier<T>, setter: Consumer<T>): Accessor<T> {
            return LambdaAccessor(getter, setter)
        }

        fun <T> of(value: T): Accessor<T> = AccessorImpl(value)
    }
}

private class AccessorImpl<T>(private var value: T) : Accessor<T> {

    override fun get(): T = value

    override fun set(t: T) {
        value = t
    }
}

private class Frozen<T>(private val su: Accessor<T>) : Accessor<T> {

    override fun get(): T = su.get()

    override fun set(t: T) {
        throw UnsupportedOperationException("frozen")
    }

    override fun frozen(): Accessor<T> = this
}

private class LambdaAccessor<T>(
    private val getter: Supplier<T>,
    private val setter: Consumer<T>,
) : Accessor<T> {
    override fun get() = getter.get()

    override fun set(t: T) = setter.accept(t)
}

private class SynchronizedAccessor<T>(val su: Accessor<T>) : Accessor<T> {
    override fun set(t: T) {
        synchronized(this) {
            su.set(t)
        }
    }

    override fun get(): T {
        synchronized(this) {
            return su.get()
        }
    }

    override fun synchronized(): Accessor<T> = this
}
